python/lsst/daf/butler/core/dimensions/sets.py

```diff
@@ -0,0 +1,558 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("DimensionSet", "DimensionNameSet", "toNameSet")
+
+from collections.abc import Set
+from collections import OrderedDict
+from .elements import DimensionElement
+
+
+def toNameSet(container):
+    """Transform an iterable of `DimensionElement` or names thereof into a
+    `DimensionNameSet`.
+    """
+    if isinstance(container, str):
+        # Friendlier error for easy mistake, e.g. union("Name") instead of
+        # union(["Name"])
+        raise TypeError("Argument must be an *iterable* over `DimensionElement` or `str`; got `str`")
+    try:
+        return DimensionNameSet(container.names)
+    except AttributeError:
+        pass
+    return DimensionNameSet(element.name if isinstance(element, DimensionElement) else element
+                            for element in container)
+
+
+class DimensionSet(Set):
+    r"""A custom set/dict hybrid class for collections of `DimensionElement`\s.
+
+    `DimensionSet` objects implement the full (immutable)
```

Comment:
Move these paragraphs to a [Notes](
https://developer.lsst.io/python/numpydoc.html#notes) section.